### PR TITLE
Remove libcheri

### PR DESCRIPTION
**Pull request by jrtc27**
This includes all the tcpdump demo code that's rather extensive and seems to have diverged in unnecessary ways.

### CODE REVIEW

File: bin/cheribsdtest/cheribsdtest.h [hunk 1]
```diff
@@ -109,17 +108,10 @@ extern struct cheribsdtest_child_state *ccsp;
 					       but not checkable */
 #define CT_FLAG_SLOW		0x00000040  /* Test is expected to take a 
 					       long time to run */
-#define	CT_FLAG_SIGNAL_UNWIND	0x00000080  /* Should fault and unwind
-					       trusted stack; checks signum
-					       and result. */
-#define	CT_FLAG_SANDBOX		0x00000100  /* Test requires that a libcheri
-					     * sandbox be created. */
 #define	CT_FLAG_SI_CODE		0x00000200  /* Check signal si_code. */
```
Comment (arichardson) [inline bin/cheribsdtest/cheribsdtest.h:111]: Should we move these values up to avoid having a gap?

**Review by bsdjhb — COMMENTED** [review]
I think you need the change to bsd.prog.mk to remove "libcheri".  I dropped that from my other review since it wasn't relevant and really should be part of this instead.

**Comment by jrtc27** [comment]
> I think you need the change to bsd.prog.mk to remove "libcheri". I dropped that from my other review since it wasn't relevant and really should be part of this instead.

Ah yeah, now done.

**Review by arichardson — APPROVED** [review]
Finally :)

File: bin/cheribsdtest/cheribsdtest.h (hunk 1 above)
Comment (jrtc27) [reply bin/cheribsdtest/cheribsdtest.h:111]: 074fbca53c7016767b7216ead624730683661fb6 already left a hole

**Review by bsdjhb — APPROVED** [review]

**Review by bsdjhb — COMMENTED** [review]
Note that this doesn't include any kernel changes, but I think that's ok.  If we want to remove kernel bits then that should probably be a separate commit(s) anyway.  I think we have mostly not included libcheri support bits in the kernel for RISC-V and Morello, and it's probably not really worth trying to GC MIPS-only bits.  I do wonder if there are any remaining comments about CHERI_PERM_SYSCALL (which I think is currently unused after this) that might need tweaking?

**Comment by bsdjhb** [comment]
Also related is that we probably need to pull the description of libcheri from various places in the ISA doc and move it into the rationale chapter as a "something we tried and this is how it worked and these are the things we liked / didn't like"

**Comment by jrtc27** [comment]
I deliberately left kernel stuff alone as it's theoretically useful for other compartmentalisation frameworks.

**Comment by jrtc27** [comment]
The checking of CHERI_PERM_SYSCALL for syscalls is in MI code (with a conf option to disable but nothing disables it). The checking for an alternate signal stack is MD code and thus the bit that's currently only implemented for MIPS.

File: bin/cheribsdtest/cheribsdtest.c [hunk 2]
```diff
@@ -721,9 +657,6 @@ main(int argc, char *argv[])
 		case 's':
 			sleep_after_test = 1;
 			break;
-		case 'u':
```
Comment (brooksdavis) [inline bin/cheribsdtest/cheribsdtest.c:724]: I wonder if we should add a temporary `errx(1, "-u is no longer supported due to libcheri removal");` incase people still have `-u`'s in CI bits.

File: bin/cheribsdtest/cheribsdtest.c [hunk 3]
```diff
@@ -769,8 +702,7 @@ main(int argc, char *argv[])
 	 * Allocate an alternative stack, required to safely process signals in
 	 * sandboxes.
 	 *
-	 * XXXRW: It is unclear if this should be done by libcheri rather than
-	 * the main program?
+	 * XXX: Is this still needed now we no longer have libcheri sandboxes?
```
Comment (brooksdavis) [inline bin/cheribsdtest/cheribsdtest.c:705]: I think this isn't needed anymore, but it doesn't hurt either.

**Review by brooksdavis — APPROVED** [review]

File: bin/cheribsdtest/cheribsdtest.c (hunk 2 above)
Comment (jrtc27) [reply bin/cheribsdtest/cheribsdtest.c:724]: I mean, we're the only ones with CheriBSD in CI, and I doubt anyone other than us has even discovered -u...

**Comment by yunji-yunji** [comment]
Is there an alternative to libcheri to write code for object capability?
I have the same need as issue #120.

To write code to examine the object capability, I have to link headers from libcheri such as `cheri_fd.h`, `cheri_invoke.h`, `cheri_memcpy.h`, etc. However, my directory `./cheri/output/rootfs-riscv64-purecap/usr/include/` from the current repository doesn't seem to include these headers.

**Comment by brooksdavis** [comment]
An object capability requires is a pair of sealed code and data capabilities with the same object type to be invoked by `CInvoke`. Ensuring safety requires compiler support or assembly to implement a calling convention which does not leak capabilities in registers across the `CInvoke` call. Unfortunately, we don't currently a compact example of this.

The headers you mention are details of the libcheri implementation and not inherently required by object capability code.

**Comment by yunji-yunji** [comment]
Thank you so much for the comment!

**Pull request merged**